benchmark/spgeam/spgeam.cpp

```diff
@@ -0,0 +1,250 @@
+#else
+using itype = gko::int32;
+const auto benchmark_name = "spgeam";
+#endif
+using Mtx = gko::matrix::Csr<etype, itype>;
+using Dense = gko::matrix::Dense<etype>;
+using mat_data = gko::matrix_data<etype, itype>;
+
+const std::map<std::string,
+               const std::function<std::shared_ptr<Mtx::strategy_type>()>>
+    strategy_map{
+        {"classical", [] { return std::make_shared<Mtx::classical>(); }},
+        {"sparselib", [] { return std::make_shared<Mtx::sparselib>(); }}};
+
+DEFINE_string(
+    strategies, "classical,sparselib",
+    "Comma-separated list of SpGEAM strategies: classical, sparselib");
+
+DEFINE_int32(
+    swap_distance, 100,
+    "Maximum distance for row swaps to avoid rows with disjoint column ranges");
+
+std::shared_ptr<Mtx> build_mtx2(std::shared_ptr<gko::Executor> exec,
+                                gko::matrix_data<etype, itype> data)
+{
+    // randomly permute n/2 rows with limited distances
+    std::vector<itype> permutation(data.size[0]);
+    std::iota(permutation.begin(), permutation.end(), 0);
+    std::default_random_engine rng(FLAGS_seed);
+    std::uniform_int_distribution<itype> start_dist(0, data.size[0] - 1);
+    std::uniform_int_distribution<itype> delta_dist(-FLAGS_swap_distance,
+                                                    FLAGS_swap_distance);
+    for (itype i = 0; i < data.size[0] / 2; ++i) {
+        auto a = start_dist(rng);
+        auto b = a + delta_dist(rng);
+        if (b >= 0 && b < data.size[0]) {
+            std::swap(permutation[a], permutation[b]);
+        }
+    }
+    for (auto &nonzero : data.nonzeros) {
+        nonzero.row = permutation[nonzero.row];
+    }
+    data.ensure_row_major_order();
+    auto result =
+        gko::share(Mtx::create(exec, data.size, data.nonzeros.size()));
+    result->read(data);
+    return result;
+}
+
+void apply_spgeam(const char *strategy_name,
+                  std::shared_ptr<gko::Executor> exec,
+                  const gko::matrix_data<etype, itype> &data, const Mtx *mtx2,
+                  rapidjson::Value &test_case,
+                  rapidjson::MemoryPoolAllocator<> &allocator)
+{
+    try {
+        add_or_set_member(test_case, strategy_name,
+                          rapidjson::Value(rapidjson::kObjectType), allocator);
+
+        auto mtx = Mtx::create(exec, data.size, data.nonzeros.size(),
+                               strategy_map.at(strategy_name)());
+        mtx->read(data);
+        auto one = gko::initialize<Dense>({1.0}, exec);
+        auto id =
+            gko::matrix::Identity<etype>::create(exec, mtx->get_size()[0]);
+        auto res = Mtx::create(exec, gko::dim<2>(mtx->get_size()));
+
+        // warm run
+        for (unsigned int i = 0; i < FLAGS_warmup; i++) {
+            res->copy_from(mtx2);
+            exec->synchronize();
+            mtx->apply(lend(one), lend(id), lend(one), lend(res));
+            exec->synchronize();
+        }
+
+        std::chrono::nanoseconds time(0);
```

Comment:
Can you not use the timer functionality that we have, for uniformity ?

Reply:
Oh yeah, these changes precede the introduction of the timer 😆